subsys/nrf_rpc/nrf_rpc_uart.c

```diff
@@ -66,47 +66,62 @@ static void work_handler(struct k_work *work)
 	uint8_t *data;
 	size_t len;
 	int ret;
-
-	len = ring_buf_get_claim(&uart_tr->rx_ringbuf, &data, NRF_RPC_MAX_FRAME_SIZE);
-
-	for (size_t i = 0; i < len; i++) {
-		if (hdlc_input_byte(uart_tr, data[i]) == 0) {
-			if (uart_tr->hdlc_state == hdlc_state_frame_found) {
-				uart_tr->receive_callback(transport, uart_tr->frame_buffer,
-							  uart_tr->frame_len, uart_tr->receive_ctx);
-				uart_tr->frame_len = 0;
-				uart_tr->hdlc_state = hdlc_state_unsync;
+	while (!ring_buf_is_empty(&uart_tr->rx_ringbuf)) {
+		len = ring_buf_get_claim(&uart_tr->rx_ringbuf, &data, NRF_RPC_MAX_FRAME_SIZE);
+		for (size_t i = 0; i < len; i++) {
+			if (hdlc_input_byte(uart_tr, data[i]) == 0) {
+				if (uart_tr->hdlc_state == hdlc_state_frame_found) {
+					uart_tr->receive_callback(transport, uart_tr->frame_buffer,
+								uart_tr->frame_len, uart_tr->receive_ctx);
+					uart_tr->frame_len = 0;
+					uart_tr->hdlc_state = hdlc_state_unsync;
+				}
 			}
 		}
-	}
 
-	ret = ring_buf_get_finish(&uart_tr->rx_ringbuf, len);
-	if (ret < 0) {
-		LOG_DBG("Cannot flush ring buffer (%d)", ret);
+		ret = ring_buf_get_finish(&uart_tr->rx_ringbuf, len);
+		if (ret < 0) {
+			LOG_DBG("Cannot flush ring buffer (%d)", ret);
+		}
 	}
 }
 
 static void serial_cb(const struct device *uart, void *user_data)
 {
 	struct nrf_rpc_uart *uart_tr = user_data;
-	int rx_len, ret;
-	uint8_t rx_buffer[8];
+	uint32_t rx_len;
+	uint8_t *rx_buffer;
+	bool new_data = false;
 
 	while (uart_irq_update(uart) && uart_irq_rx_ready(uart)) {
-		rx_len = uart_fifo_read(uart, rx_buffer, sizeof(rx_buffer));
-		if (rx_len <= 0) {
-			continue;
-		}
+		new_data = false;
+		rx_len = ring_buf_put_claim(
+			&uart_tr->rx_ringbuf, &rx_buffer,
+			uart_tr->rx_ringbuf.size);
+		if (rx_len > 0) {
+			rx_len = uart_fifo_read(uart, rx_buffer, rx_len);
+			int err = ring_buf_put_finish(
+				&uart_tr->rx_ringbuf, rx_len);
+			if (err != 0) {
+				__ASSERT_NO_MSG(err == 0);
+			}
+			if (rx_len <= 0) {
+				continue;
+			} else {
+				new_data = true;
+			}
+		} else {
+			uint8_t dummy;
 
-		ret = ring_buf_put(&uart_tr->rx_ringbuf, rx_buffer, (uint32_t)rx_len);
-		if (ret < rx_len) {
-			LOG_ERR("Rx buffer doesn't have enough space. "
-				"Bytes pending: %d, written: %d",
-				rx_len, ret);
-			break;
+			/* No space in the ring buffer - consume byte. */
+			LOG_WRN("RX ring buffer full.");
+
+			rx_len = uart_fifo_read(uart, &dummy, 1);
 		}
 
-		k_work_submit_to_queue(&uart_tr->rx_workq, &uart_tr->cb_work);
+		if (new_data) {
```

Comment:
nit: You might call this after the loop so that `k_work` is always scheduled only once per ISR.

Reply:
fixed.